### DISCUSSION TITLE
Access to global variable in fn

### DISCUSSION THREAD

**Discussion by gryznar** [modular/modular#448] [Q&A]
Last update added support for global variables declaration. So I wrote some code to test that:
```mojo
let x: UInt8 = 1


fn test_global():
    print(x)

test_global()
```

This gives me an error about unknown declaration "x". Is this not implemented yet or am I doing sth wrong?

**Comment by lattner** [comment]
@Mogball ?

**Reply by Mogball** [reply]
Global variables were added to the language but they have not been wired into the REPL environment yet. The REPL environment layers extra features on the language to provide redefinition and top-level variables, and using global variables to enable the code you wrote has not occurred yet. Sorry for the confusion!

**Reply by gryznar** [reply]
If there is a differentiation for REPL, so is that mean that Mojo can behave also like a classical interpreted / scripting language with almost no startup time (like a Python 😁)? 

Sorry maybe stupid question, but as I have heard (I am a begginer in compiler related stuff) JIT is not a magic bullet for every usecase

**Reply by Mogball** [reply]
> so is that mean that Mojo can behave also like a classical interpreted / scripting language

Yes. The goal of the Mojo REPL is to provide a development experience like Python and other interpreted languages. It pairs especially well with Jupyter (effectively a frontend for the REPL).

> with almost no startup time 

As you have observed, using a JIT is not a magic bullet for making a compiled language behave like a truly interpreted one. The cost of runtime speed is paid in compilation time, and in the REPL that happens for each state. That said, Mojo does have an interpreter! It is used to evaluate Mojo code at compile time, but isn't something directly accessible at the moment.

**Reply by gryznar** [reply]
@Mogball So maybe it is worth to point it out in documentation and language description that Mojo can behave also like a interpreted one and replicate Python behavior?  It looks like a real game changer! Hiding this feature may confuse and misleads people.

One more thing to add is also mentioning that (as of my knowledge) Mojo will have optional GIL to handle C extensions for example (another great advantage). It may be another game changer, because all current extensions could work out-of-the-box without any changes required and hopefully with the same speed.

Currently it may seem that use cases are somehow restricted, because of not clarifying that details

**Reply by Mogball** [reply]
The feature is accessible in the sense that compile-time expressions are evaluated using an interpreter. Like

```python
fn sum(lb: Int, ub: Int, step: Int):
  var total = 0
  for i in range(lb, ub, step):
    total += i
  return total

fn main():
  alias x = sum(10, 20, 2) # `x` is evaluated at compile time: the interpreter invokes `sum`
  print(x) # should compile to `print(70)`
```

By "it is not directly accessible", I mean that it is not possible to run normal Mojo code using the interpreter. It is only invoked during compilation.

**Reply by gryznar** [reply]
Are there plans to make it accessible? Or interpreter is tend to stay "implementation detail"?

**Reply by Mogball** [reply]
> Are there plans to make it accessible?

Not at the moment. The interpreter is a low-level affair so some amount of "compilation" would still have to occur before it can be invoked.

**Reply by gryznar** [reply]
Make sense! Thanks for clarification :)
